src/reflect/scala/reflect/internal/util/ReusableInstance.scala

```diff
@@ -15,20 +15,34 @@ package reflect
 package internal
 package util
 
-/** A wrapper for a re-entrant, cached instance of a value of type `T`.
+import scala.collection.mutable.ArrayBuffer
+
+/** A wrapper for a list of cached instances of a value of type `T`.
+  * The wrapper is recursion-reentrant: several instances are kept, so
+  * at each depth of reentrance we are reusing the instance for that.
+  *
+  * An instance is created upon creating this object, and more instances
+  * are allocated dynamically, on demand, when reentrance occurs.
   *
   * Not thread safe.
   */
 final class ReusableInstance[T <: AnyRef](make: () => T, enabled: Boolean) {
-  private val cached = make()
-  private var taken = false
+  private[this] val cached = new ArrayBuffer[T]
```

Comment:
Using `ArrayStack` would make the code a little cleaner (just use `push`/ `pop` rather than `+=` `apply(size-1)`.

Reply:
Oh, but I guess it wouldn't ever reuse items that you `pop` off..

Reply:
does this need to be cleared at the end of the compile run?

Reply:
I cannot see why would it need to. The principle is the same as before the `ReusableInstance`: instead of creating an immutable object, we have mutable instances that, wherever they are to be used, have to be reset first.

Reply:
I noticed it pre-allocates if not enabled. I was curious to experiment, so I added a commit at https://github.com/scala/scala/pull/8871 but I was too lazy to fetch this commit. The other change was to use the companion apply.

Reply:
It could be called `ReusableInstances` now. I guess `Pool` is old-fashioned.

Reply:
> I noticed it pre-allocates if not enabled

We could make it to initialise as `null` and allocate upon first usage. However, given that `ReusableInstance` objects are a sort of static variable of the compiler, and that there are four of five of them, this should not have much impact.

Reply:
On the other PR, I noted there are hundreds for sym lookup. The "not enabled" case is for reflect, I guess, so I don't know what the usage profile is.